src/mpid/ch4/netmod/ofi/ofi_send.h

```diff
@@ -581,28 +521,97 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
     void *send_buf = MPIR_get_contig_ptr(buf, dt_true_lb);
     MPIR_GPU_query_pointer_attr(send_buf, &attr);
 
-    if (MPIR_CVAR_CH4_OFI_ENABLE_INJECT && !syncflag && dt_contig &&
+    bool is_gpu = false;
+    bool need_pack = false;
+    bool need_mr = false;
+    bool do_inject = false;
+    bool do_iov = false;
+    bool do_gpu_pipelining = false;
+    bool do_striping = false;
+    bool do_huge = false;
+
+    /* check gpu */
+    if (data_sz && MPL_gpu_query_pointer_is_dev(send_buf, &attr)) {
+        is_gpu = true;
+        MPIDI_OFI_register_am_bufs();
+    }
+
+    /* check gpu paths */
+    if (is_gpu) {
+        if (!MPIDI_OFI_ENABLE_HMEM) {
+            need_pack = true;
+        } else if (attr.type != MPL_GPU_POINTER_DEV) {
+            /* non-strict gpu ptr (ZE shared host) */
+            need_pack = true;
```

Comment:
I'm not sure if this restriction is needed for other backends. It won't be incorrect, but might be reduced performance in case they don't need packing in this scenario

Reply:
I think currently only ZE `MPL_gpu_query_pointer_is_dev` can be different from `attr.type != MPL_GPU_POINTER_DEV`. But if other backend start to have similar requirement, the `MPL_gpu_query_pointer_is_dev` should accommodate it.